pinot-core/src/main/java/org/apache/pinot/core/data/table/Table.java

```diff
@@ -0,0 +1,46 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.table;
+
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.apache.pinot.common.request.AggregationInfo;
+import org.apache.pinot.common.request.SelectionSort;
+import org.apache.pinot.common.utils.DataSchema;
+
+
+/**
+ * Table of data records to pass data between operators
+ */
+public interface Table {
```

Comment:
As discussed offline, there are three options to consider, which one is the PR going to evolve into?
1. Have separate api's on this interface for upsert/append and let caller be responsible for using the correct one. There will be one implementation that supports all.
2. Have separate implementations for upsert/append use cases (to support selection, group-by, order-by).
3. This interface provides a raw data structure and caller side has the complexity to deal with selection/group-by/order-by.

Reply:
For now, let's just do one implementation that supports group by + order by. I don't see a lot of value addition in using this for aggregation and selection as of now.